corehq/util/couch.py

```
@@ -33,3 +36,130 @@ def categorize_bulk_save_errors(error):
+
+    iter_update returns an object with the following properties:
+    saved_ids, deleted_ids, ignored_ids,
+
+    Ex: mark dimagi users as cool, delete the Canadians, and ignore the rest
+
+        def mark_cool(user_dict):
+            user = CommCareUser.wrap(user_dict)
+            if user.is_dimagi:
+                user.is_cool = True
+                return user
+            if user.language == "Canadian":
+                return "DELETE"
+
+        iter_update(CommCareUser.get_db(), mark_cool, all_user_ids)
+
+    This looks up and saves docs in chunks and is intended for large changes
+    such as a migration.  If an id is not found, it is skipped.  In the
+    event of a BulkSaveError, it will re-process the unsaved documents.
+    Wrapping is optional, and this function will unwrap if needed.
+    """
+    IterResults = namedtuple('IterResults', ['ignored_ids', 'not_found_ids',
+                                             'deleted_ids', 'updated_ids'])
+    results = IterResults(set(), set(), set(), set())
+
+    def _iter_update(doc_ids):
+        with IterDB(db, chunksize=100) as iter_db:
+            for chunk in chunked(set(doc_ids), 100):
+                for res in get_docs(db, keys=chunk):
+                    raw_doc = res.get('doc')
+                    doc_id = res.get('id', None)
+                    if not raw_doc or not doc_id:
+                        results.not_found_ids.add(res['key'])
+                    else:
+                        doc = fn(raw_doc)
+                        if isinstance(doc, dict):
+                            iter_db.save(doc)
+                        elif doc == "DELETE":
+                            iter_db.delete(raw_doc)
+                        else:
+                            results.ignored_ids.add(doc_id)
+
+        results.updated_ids.update(iter_db.saved_ids)
+        results.deleted_ids.update(iter_db.deleted_ids)
+
+        if iter_db.error_ids:
+            _iter_update(iter_db.error_ids)
```

yeah, agree with @benrudolph and @snopoke. feels like error handling should be the caller's responsibility to think through and the default should be to fail hard.